api/agent/agent.go

```diff
@@ -737,114 +745,183 @@ func newSizerRespWriter(max uint64, rw http.ResponseWriter) http.ResponseWriter
+		lastState := state.GetState()
+		state.UpdateState(ctx, ContainerStateDone, call.slots)
 
-	cookie, err := a.driver.CreateCookie(ctx, container)
-	if err != nil {
-		call.slots.queueSlot(&hotSlot{done: make(chan struct{}), fatalErr: err})
+		tok.Close() // release cpu/mem
+
+		// IMPORTANT: evict token is deleted *after* resource token.
+		// This ordering allows resource token to be freed first, which means once evict token
+		// is deleted, eviction is considered to be completed.
+		a.evictor.DeleteEvictToken(evictor)
+
+		statsUtilization(ctx, a.resources.GetUtilization())
+		if evictor.isEvicted() {
+			logger.Debugf("Hot function evicted")
+			statsContainerEvicted(ctx, lastState)
+		}
+	}()
+
+	go func() {
+		isInitialized, err := a.waitInitialize(ctx, caller, evictor, udsWait)
+		if err != nil {
+			tryQueueErr(err, errQueue)
+		}
+		if isInitialized {
+			close(initialized)
+		} else {
+			cancel()
+		}
+	}()
+
+	container, err = newHotContainer(ctx, call, &a.cfg, id, udsWait)
+	if tryQueueErr(err, errQueue) != nil {
 		return
 	}
 
-	// WARNING: we wait forever.
-	defer cookie.Close(common.BackgroundContext(ctx))
+	cookie, err = a.driver.CreateCookie(ctx, container)
+	if tryQueueErr(err, errQueue) != nil {
+		return
+	}
 
-	shouldPull, err := cookie.ValidateImage(ctx)
-	if err != nil {
-		call.slots.queueSlot(&hotSlot{done: make(chan struct{}), fatalErr: err})
+	needsPull, err := cookie.ValidateImage(ctx)
+	if tryQueueErr(err, errQueue) != nil {
 		return
 	}
 
-	if shouldPull {
+	if needsPull {
+		ctx, cancel := context.WithTimeout(ctx, a.cfg.HotPullTimeout)
 		err = cookie.PullImage(ctx)
-		if err != nil {
-			call.slots.queueSlot(&hotSlot{done: make(chan struct{}), fatalErr: err})
+		cancel()
+		if ctx.Err() == context.DeadlineExceeded {
+			err = models.ErrDockerPullTimeout
+		}
+		if tryQueueErr(err, errQueue) != nil {
 			return
 		}
 	}
 
 	err = cookie.CreateContainer(ctx)
-	if err != nil {
-		call.slots.queueSlot(&hotSlot{done: make(chan struct{}), fatalErr: err})
+	if tryQueueErr(err, errQueue) != nil {
 		return
 	}
 
-	cancel()
-	ctx, shutdownContainer := context.WithCancel(common.BackgroundContext(ctx))
-	defer shutdownContainer() // close this if our waiter returns, to call off slots, needs to follow cookie.Close so the cookie crumbles
-
-	udsAwait := make(chan error)
-	// start our listener before starting the container, so we don't miss the pretty things whispered in our ears
-	// make sure this thread has the shutdownContainer context in case the container exits
-	go inotifyUDS(ctx, container.UDSAgentPath(), udsAwait)
-
-	waiter, err := cookie.Run(ctx)
-	if err != nil {
-		call.slots.queueSlot(&hotSlot{done: make(chan struct{}), fatalErr: err})
+	waiter, err = cookie.Run(ctx)
+	if tryQueueErr(err, errQueue) != nil {
 		return
 	}
 
 	go func() {
-		defer shutdownContainer() // also close if we get an agent shutdown / idle timeout
+		defer cancel() // also close if we get an agent shutdown / idle timeout
```

Comment:
def found the last name more obvious

Reply:
maybe, but it was kind of misleading. Which container? "container" from newHotContainer? docker container in cookie? Here cancel is generic, and it means "we cancel all operations" inside this function. This PR moves the cancel to the beginning of runHot(), which covers a lot more than it did before.